### ISSUE TITLE
Support CORS fully with pre-flight requests using HTTP OPTIONS and server-side config for CORS

### ISSUE THREAD

**Issue by hodgesrm** [ClickHouse/ClickHouse#18693]
**Use case**
ClickHouse cannot accept connections from web browser-based query tools that implement [CORS protocol preflight requests](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request).  The reason is preflight checks use an OPTIONS request, which ClickHouse does not implement in the HTTP interface. 

The problem occurs in tools like Grafana and [ObservableHQ](https://observablehq.com), whenever (a) ClickHouse is on a different server from the web page source and (b) the call to ClickHouse is processed directly in the browser. This use case occurs commonly when integrating ClickHouse cloud implementations with SaaS-based BI tools. In each case the browser will do the following: 

1. Send an OPTIONS request to check if ClickHouse allows cross-site requests.  The browser expects to see an Access-Control-Allow-Origin header that whitelists the original site and a 200 status code.  
2. Once the OPTIONS request returns successfully, send the actual request. 

When Grafana tries to do this it fails with the following message in the browser console. ObservableHQ fails similarly. 

```
Access to fetch at 'https://github.demo.trial.altinity.cloud:8443/?add_http_cors_header=1' from origin 'http://localhost:3000' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
```
In the case of Grafana you can work-around CORS problems by running the ClickHouse queries on the Grafana server (set Access to 'Server' in the Datasource definition).  For ObservableHQ this is a blocking bug.  

There is another problem in that the current add_http_cors_header=1 option used to allow ClickHouse to return a Access-Control-Allow-Origin header makes client URLs more cumbersome. There's also a security issue in that CORS should be enabled by the server, not triggered by the client. Malicious sites can undo CORS protection by changing the URL. 

**Describe the solution you'd like**
ClickHouse should implement [server-side CORS support](https://enable-cors.org/server.html) fully, which includes the following: 

1. Implement OPTIONS requests for CORS preflight checks.  The expected behavior is shown in the referenced page above and illustrated in the attached Python3 script showing a typical OPTIONS request for pre-flight checks. Note that you can run a preflight request without authenticating. 
2. Make CORS support a server-side setting.  It would make sense to enable cors using a regex with '*' to designate accept all sites.  CORS should be off by default. 
3. Deprecate the add_http_cors_header=1 URL option as soon as possible. 

**Describe alternatives you've considered**
An alternative option is to enable CORS using a proxy, but this seems cumbersome and hard for developers.  If ClickHouse supports HTTP it should do it completely, including CORS. 

**Additional context**
As noted above this is a much more prominent problem with ClickHouse cloud platforms.

**Comment by hodgesrm** [comment]
Sample script showing typical OPTIONS request for CORS preflight check. 
[
[options-request.py.txt](https://github.com/ClickHouse/ClickHouse/files/5761362/options-request.py.txt)
](url)

**Comment by filimonov** [comment]
Yep, it seems preflight CORS just hangs waiting to request body

That hangs:
```
curl --verbose -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://localhost" 'http://localhost:8123/?add_http_cors_header=1&query=SELECT%20*%20FROM%20numbers(10)'
*   Trying 127.0.0.1:8123...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8123 (#0)
> OPTIONS /?add_http_cors_header=1&query=SELECT%20*%20FROM%20numbers(10) HTTP/1.1
> Host: localhost:8123
> User-Agent: curl/7.68.0
> Accept: */*
> Access-Control-Request-Method: GET
> Origin: http://localhost
> 
^C
```

```
echo '' | curl --verbose -X OPTIONS -H "Access-Control-Request-Method: GET" -H "Origin: http://localhost" 'http://localhost:8123/?add_http_cors_header=1&query=SELECT%20*%20FROM%20numbers(10)' --data-binary @-
*   Trying 127.0.0.1:8123...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8123 (#0)
> OPTIONS /?add_http_cors_header=1&query=SELECT%20*%20FROM%20numbers(10) HTTP/1.1
> Host: localhost:8123
> User-Agent: curl/7.68.0
> Accept: */*
> Access-Control-Request-Method: GET
> Origin: http://localhost
> Content-Length: 1
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 1 out of 1 bytes
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Date: Mon, 04 Jan 2021 15:58:40 GMT
< Connection: Keep-Alive
< Content-Type: text/tab-separated-values; charset=UTF-8
< X-ClickHouse-Server-Display-Name: laptop-5591
< Transfer-Encoding: chunked
< X-ClickHouse-Query-Id: b9c1e486-414c-4217-9c05-ba2ae19fd50d
< X-ClickHouse-Format: TabSeparated
< X-ClickHouse-Timezone: Europe/Warsaw
< Access-Control-Allow-Origin: *
< Keep-Alive: timeout=3
< X-ClickHouse-Summary: {"read_rows":"0","read_bytes":"0","written_rows":"0","written_bytes":"0","total_rows_to_read":"0"}
< 
0
1
2
3
4
5
6
7
8
9
* Connection #0 to host localhost left intact
```

**Comment by hodgesrm** [comment]
Yep, that's the behavior I get as well for curl. Internally it seems ClickHouse might be treating OPTIONS as a POST and trying to read the body. The behavior actually depends on how the client forms the request. In the Python sample it just fails. 

I've implemented CORS previously in Java.  It's not hard.  I might try doing the PR on this one myself.

**Comment by alexey-milovidov** [comment]
Bump. Still relevant.
